Paginate AWS results, submit total GB usage and volume count to appropriate statsum series

rstiyer opened the pull request:
### Code Changes

The following code relies on @SamanthaYu's ongoing work on calculating the total GB usage and volume count in each region, separated by volume type and status (see [https://github.com/rstiyer/ec2-manager/pull/1](url)).

First of all, I modified the AWS API call to request paginated data. The max page size is 500, and I don't see any reason not to use the maximum size. The 'describeVolumes' endpoint appears to return all data if you don't specify a page size, but we don't want to rely on this since many other endpoints only return the max page size.

Second, I changed our approach with `_handleVolumeData(...)`. Previously, we called the function for every individual volume, but that doesn't make very much sense since we really care about the aggregation of all volumes within the region. Instead, I changed its parameter to be the list of all volumes in the region. This function now calls our helper functions for calculating and updating the totals per region, `_calculateVolumeTotals` and `_updateVolumeTotals`. I added 'volume' to the name of these functions to differentiate these functions from those that will eventually be added to complete similar functionality for the instance and spot request endpoints. I thought it made sense to have these helper calls in `_handleVolumeData` to encourage the single-responsibility principle.

Finally, I created a preliminary implementation of `_updateVolumeTotals`. This takes one parameter: an object containing the required values returned by `_calculateVolumeTotals`. It extracts the relevant information and then updates the count of appropriate statsum series. I added a bunch of asset statements (which are now commented out) to ensure that incorrect data is not accidentally submitted to the statsum series.

### Questions:
* It may take more than 90 seconds for `_sweepVolumes` to complete for the real data, which means that the 'global' updates might not all be aggregated together in the same 5 minute period. We can't test this ourselves, unfortunately. If this is indeed the case, we'll have to store the overall totals and update the 'global' series after processing every region.

rstiyer commented:
For some reason, I can't tag anyone as a reviewer even though this is my own repo.

@imbstack @jhford @SamanthaYu

imbstack commented on lib/housekeeping.js:
```diff
@@ -38,17 +38,109 @@ class HouseKeeper {
+    let region =  Object.keys(totalsByCategory)[0];
+    
+    let volumeTypes = totalsByCategory[region];
+    //assert(typeof volumeTypes === 'object');
```
I assume these `assert`s are leftover debugging things? We can remove them before this is merged. If they're useful to keep around feel free to leave the asserts in, but no need to comment them out in that case!

rstiyer replied:
That's something I wanted to ask about, actually. I commented them out for now primarily so the PR would pass the tests. I was thinking that it made sense to include these assert statements as a way to guarantee that the right data is submitted to all statsum series, but does it make sense to include these past the testing phase?

imbstack commented on lib/housekeeping.js:
```diff
@@ -38,17 +38,109 @@ class HouseKeeper {
+
+      //assert(data.hasOwnProperty('active'));
+      //assert(data.hasOwnProperty('unused'));
+      let active = data.active;
```
This can be made a bit nicer with 

`let {active, unused} = data;`

if you'd like!

rstiyer replied:
I'd forgotten about that syntax, and it's definitely cleaner! Thanks!

imbstack commented on lib/housekeeping.js:
```diff
@@ -38,17 +38,109 @@ class HouseKeeper {
+      //assert(active.hasOwnProperty('gb'));
+      //assert(active.hasOwnProperty('count'));
+
+      this.monitor.count('ebs-volumes:${region}:gb-usage', active.gb);
```
Forgive me if this is me missing the point, but do these need to be separated on `volumeType` as well? It seems like what you have on line 101 should be up here maybe?

rstiyer replied:
This is something to ask @jhford about. In the UCOSP project document, John listed just this set of series as a starting point. It might make sense to add a couple more to distinguish between active and unused volumes of each type, as well.

imbstack commented on lib/housekeeping.js:
```diff
@@ -38,17 +38,109 @@ class HouseKeeper {
+
+      let totalGbUsage = active.gb + unused.gb;
+      let totalCount = active.count + unused.count;
+      this.monitor.count('ebs-volumes:global:gb-usage', totalGbUsage);
```
And similarly, should this totalGbUsage report happen outside the forEach loop?

rstiyer replied:
You're very right, that makes a lot more sense.

I'll still need to rely on the following note from the statsum repo, which leads to the question I mentioned in the PR comments:
> As a rule of thumb, it is reasonable to accumulate metrics for 30 to 90 seconds before flushing. As statsum will only reports every 5min.

imbstack reviewed (commented):
Looking good so far!

SamanthaYu commented on lib/housekeeping.js:
```diff
@@ -38,17 +38,109 @@ class HouseKeeper {
+    // by specifying NextToken in a subsequent request.
+    let hasNextToken = true;
+    let nextToken = null;
+    while (hasNextToken) {
```
Maybe get rid of `hasNextToken` and change this loop to a do/while with the condition `nextToken != null`.

rstiyer replied:
That's a great idea! I forgot that such a thing existed.

SamanthaYu replied:
I don't think every language has do/while loops so it's not very popular. In "The Art of Readable Code", they compare it to going down a ski slope with the warning sign at the bottom (https://www.safaribooksonline.com/library/view/the-art-of/9781449318482/ch07s04.html).

rstiyer commented:
@SamanthaYu: do you think this approach of using `_handleVolumeData(...)` makes sense?

SamanthaYu commented:
I discussed your PR with @djmitche and he suggested this alternative using callbacks:
```
async forEachVolume(callback) {
     /* call AWS for each page, invoking callback (vol) for each volume in each page */
}
async sweepVolumes() {
     let totalGb = 0;
     let totalVolumes = 0;

     await forEachVolume((vol) => {
          totalGb += vol.gb;
          totalVolumes += 1;
     }
}
```
Your approach makes sense though so I would still stick with what you have.

SamanthaYu commented on lib/housekeeping.js:
```diff
@@ -38,17 +38,106 @@ class HouseKeeper {
+          Name: 'status',
+          Values: ['available', 'in-use'],
+        }],
+        MaxResults: 500, // This is the maximum page size allowed by the describeVolumes endpoint.
```
@rstiyer Thanks for figuring out the pagination!

djmitche commented:
We do something similar [here](https://github.com/taskcluster/aws-provisioner/blob/e6efa6c637deb9a6199de6cbde31eed1a99292d7/src/api-v1.js#L123-L126) -- although much simpler (just pushing things onto a list).  But I only brought that up as an alternative, NOT to say that you need to change from your current implementation!

SamanthaYu commented on lib/housekeeping.js:
```diff
@@ -58,53 +60,66 @@ class HouseKeeper {
 
-    volumeTypes.forEach(function(data) {
-      //assert(typeof data === 'object');
+      volumeTypes.forEach(function(data) {
```
I think `volumeTypes` is an object. If it is, we're going to have to traverse through by keys:
`Object.keys(volumeTypes).foreach(function(type) {}`

I made a possible implementation in https://github.com/rstiyer/ec2-manager/pull/3 with some changes to what `active` and `unused` are initialized as.

rstiyer replied:
If you look at my assumed format, `volumeTypes` is in fact a list of objects where the key of each object is the type and the value are the GB and count totals among volumes of that type in the given region.

rstiyer replied:
I don't think there is any difference at all, to be honest. I just went with this approach because we don't know exactly how many different volume types we can expect and it's more natural to loop over a list of objects rather than the keys of the parent object.

SamanthaYu replied:
I'm sorry I was getting confused between the brackets and didn't realize that it was a list of objects. Your implementation makes sense then. I agree that it is more natural to loop over a list of objects. However, I'm not a huge fan of how we would have to access the elements: e.g. `let type = Object.keys(data)[0];` is not that readable compared to just getting the `type` as the key in the `.foreach()` loop.

rstiyer replied:
I think your implementation is probably better! :) Definitely clearer, so let's go for it!

SamanthaYu replied:
Awesome. How about we merge your PR into `calculate-and-update-totals`? We can merge my PR with this new implementation in afterwards. We're still not merging it into the canon repo so it's OK that @imbstack and @djmitche hasn't had a chance to look at your most recent changes.

Or we could also just copy/paste my changes to `_submitVolumeTotals()` into your PR then merge it in later.

SamanthaYu replied:
Hopefully, the PR for merging `calculate-and-update-totals` into the canon repo wouldn't be too big.

SamanthaYu commented on lib/housekeeping.js:
```diff
@@ -58,53 +60,66 @@ class HouseKeeper {
+        //assert(active.hasOwnProperty('gb'));
+        //assert(active.hasOwnProperty('count'));
+
+        this.monitor.count('ebs-volumes:${region}:gb-usage', active.gb);
```
`this` appears to be undefined in `_submitVolumeTotals()`. When I get rid of `this` in `this.monitor.count()`, my tests in https://github.com/rstiyer/ec2-manager/pull/3 are able to run. I'm guessing it's because `_submitVolumeTotals()` is called by `_handleVolumeTotals()` as `this._submitVolumeTotals()`.

The `this` objects in JS always get me confused so that may or may not be the right solution.

rstiyer replied:
It's interesting that getting rid of `this` makes it work, since we call all functions with `this` (e.g. `sweep` calls `this._handleVolumeData` which in turn calls `this._submitVolumeTotals`). I guess I'll need to look into this.

SamanthaYu replied:
It's also weird that we don't need to get rid of the `this` on lines 116 and 117. Maybe it has something more to do with the nested loops.

SamanthaYu replied:
I changed the function in `foreach()` to an arrow function (as per @djmitche's suggestions), and the `this` object is now defined! I also changed it in `_calculateVolumeTotals()` in this commit: https://github.com/SamanthaYu/ec2-manager/commit/5df323c7181b1544d3bfad348bd5c35d105009f9

SamanthaYu commented on test/housekeeping_test.js:
```diff
@@ -333,55 +334,48 @@ describe('House Keeper', () => {
+        AvailabilityZone: 'us-west-2', 
+        CreateTime: new Date().toString(), 
+        Size: 8, 
+        SnapshotId: "snap-1234567890abcdef0", 
```
nit: SnapshotId and VolumeId should be unique to each volume

rstiyer replied:
Yeah, weird. I fixed that in one spot but it seems to have gotten mixed up during the branching and merging..

rstiyer committed:
Use 'measure' rather than 'count' to track metrics, define prefix 'ebs-volumes' to make things look nicer, replace ':' with '.' for consistency

Also, change ids in test so they are unique.

rstiyer committed:
Add 'type' qualifier to statsum metric name for clarity.

rstiyer committed:
Resolve merge conflicts in 'lib/housekeeping.js' and 'test/housekeeping_test.js' after merging @SamanthaYu's changes

rstiyer committed:
Add assert statements back to _submitVolumeTotals

Since it isn't possible to programatically check that the statsum series are being updated correctly, use assert statements to verify the input is formatted as expected.
